newIDE/app/src/Profile/UsernameField.js

```
@@ -2,42 +2,141 @@
 
+const usernameAvailabilityErrorMessage =
+  'This username is already used, please pick another one.';
+
 export const UsernameField = ({
+  initialUsername,
   value,
   onChange,
   errorText,
   allowEmpty,
+  onAvailabilityChecked,
+  onAvailabilityCheckLoading,
 }: Props) => {
-  const getUsernameErrorText = () =>
-    isUsernameValid(value, allowEmpty) ? undefined : usernameFormatError;
+  const usernameFormattingError = React.useMemo(
+    () =>
+      isUsernameValid(value, { allowEmpty: !!allowEmpty })
+        ? undefined
+        : usernameFormatErrorMessage,
+    [value, allowEmpty]
+  );
+
+  const [
+    usernameAvailabilityError,
+    setUsernameAvailabilityError,
+  ] = React.useState<?string>(null);
+
+  const updateUsernameAvailability = ({
+    usernameAvailability,
+    error,
+  }: {
+    usernameAvailability: UsernameAvailability | null,
+    error: string | null,
+  }) => {
+    onAvailabilityChecked(usernameAvailability);
+    setUsernameAvailabilityError(error);
+    onAvailabilityCheckLoading(false);
+  };
+
+  const checkUsernameAvailability = useDebounce(async (username: string) => {
+    // If just casing change, the username is always available.
+    if (
+      initialUsername &&
+      initialUsername.toLowerCase() === username.toLowerCase()
+    ) {
+      updateUsernameAvailability({
+        usernameAvailability: { username, isAvailable: true },
+        error: null,
+      });
+      return;
+    }
+    // no username or invalid, no need to check availability.
+    if (!username || !isUsernameValid(username, { allowEmpty: !!allowEmpty })) {
+      updateUsernameAvailability({
+        usernameAvailability: null,
+        error: null,
+      });
+      return;
+    }
+
+    try {
+      const usernameAvailability = await getUsernameAvailability(username);
+
+      if (!usernameAvailability) {
+        console.error('Unable to check username availability.');
```

This is the same code as in the catch block, you could throw an error here to avoid repetition